server/blob/blob-deduplicating/src/main/scala/org/apache/james/server/blob/deduplication/GC.scala

```diff
@@ -0,0 +1,185 @@
+   */
+  def collectibles(targetedGeneration: Generation): Set[Generation] =
+    Generation.range(this, targetedGeneration.previous(GC.temporization)).toSet
+
+  def <(that: Generation): Boolean = compareTo(that) < 0
+  def <=(that: Generation): Boolean = compareTo(that) <= 0
+  def >(that: Generation): Boolean = compareTo(that) > 0
+  def >=(that: Generation): Boolean = compareTo(that) >= 0
+}
+
+object Generation {
+  val first: Generation = apply(0)
+
+  def apply(id: Long): Generation = {
+    if (id < 0) {
+      NonExistingGeneration
+    } else {
+      ValidGeneration(id)
+    }
+  }
+
+  def range(start: Generation, end: Generation): Seq[Generation] = (start, end) match {
+    case (NonExistingGeneration, NonExistingGeneration) => Seq(NonExistingGeneration)
+    case (ValidGeneration(_), NonExistingGeneration) => Nil
+    case (NonExistingGeneration, ValidGeneration(id)) =>  NonExistingGeneration +: (0L to id).map(Generation.apply)
+    case (ValidGeneration(id1), ValidGeneration(id2)) => (id1 to id2).map(Generation.apply)
+  }
+}
+
+/**
+ * Generation which has existed
+ */
+case class ValidGeneration(id: Long) extends Generation {
+  override def previous: Generation = Generation(id - 1)
+
+  override def next: Generation = copy(id + 1)
+
+  override def compareTo(t: Generation): Int = t match {
+    case NonExistingGeneration => 1
+    case that: ValidGeneration => id.compareTo(that.id)
+  }
+
+}
+
+/**
+ * NullObject for the initialisation of the GC
+ */
+case object NonExistingGeneration extends Generation {
+  override def previous: Generation = NonExistingGeneration
+
+  override def next: Generation = Generation.first
+
+  override def compareTo(t: Generation): Int = t match {
+    case NonExistingGeneration => 0
+    case _: ValidGeneration => -1
+  }
+}
+
+/**
+ * A run of the GC regarding a Set of Generations
+ */
+case class Iteration(id: Long, processedGenerations: Set[Generation], lastGeneration: Generation) {
+  def next(generations: Set[Generation], lastGeneration: Generation): Iteration = Iteration(id + 1, generations, lastGeneration)
+}
+
+object Iteration {
+  def initial: Iteration = Iteration(0, Set(), NonExistingGeneration)
+}
+
+case class ExternalID(id: String) // TODO
+
+/**
+ * Modelized users' interactions related to blobs
+ */
+sealed trait Event {
+  def blob: BlobId
+  def externalId: ExternalID
+  def generation: Generation
+}
+
+case class Reference(externalId: ExternalID, blobId: BlobId, generation: Generation) extends Event {
+  override def blob: BlobId = blobId
+}
+
+case class Deletion(generation: Generation, reference: Reference) extends Event {
+  override def blob: BlobId = reference.blob
+  override def externalId: ExternalID = reference.externalId
+}
+
+object Events {
+  def getLastGeneration(events: Seq[Event]): Generation = events.map(_.generation).maxOption
+    .getOrElse(Generation.first)
+
+}
+
+case class Report(iteration: Iteration, blobsToDelete: Set[(Generation, BlobId)])
```

Comment:
Given that Generation should be an intrinsec property of the BlobId for a deduplicatng BlobStore (and related context) this can become IMO `case class Report(iteration: Iteration, blobsToDelete: Set[BlobId])`

(Which makes sense as a BlobId belongs to a single generation as far as I am aware of)

Reply:
Sure, but see https://github.com/linagora/james-project/pull/3220#discussion_r422919815

Reply:
> no, and it will not in that PR since it's related to the BlobStore work

Well my above comment shows it actually impact the way the GC algorithm is built (or at least expressed)

That's not a hard blocker for me but I would find it nicer.